halo/app/monitor.go

```diff
@@ -167,3 +184,44 @@ func dirSize(path string) (int64, error) {
 
 	return size, nil
 }
+
+// startMonitoringAPI registers metrics and health endpoints.
+func startMonitoringAPI(ctx context.Context, cfg *cmtcfg.Config, cmtNode *node.Node, syncAbort chan error, status *readinessStatus) {
+	mux := http.NewServeMux()
+	mux.Handle("/", promhttp.Handler())
+
+	// On the `/ready` endpoint, we serve the readiness of the halo node.
+	// Additionally, in case when the node is not ready, the response status code is set to 503.
+	mux.HandleFunc("/ready", func(w http.ResponseWriter, _ *http.Request) {
+		w.Header().Set("Content-Type", "application/json")
+
+		if !status.ready() {
+			w.WriteHeader(http.StatusServiceUnavailable)
+		}
+		if err := status.serialize(w); err != nil {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+	})
+
+	server := &http.Server{
+		Addr:              cfg.Instrumentation.PrometheusListenAddr,
+		ReadHeaderTimeout: 3 * time.Second,
+		Handler:           mux,
+	}
+
+	go func() {
+		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
+			errorMsg := fmt.Sprintf("HTTP server failed to start: %v", err)
+			syncAbort <- errors.New(errorMsg)
+		}
+	}()
+
+	go func() {
```

Comment:
rather return a shutdown function and call it in main `Run` shutdown func.

In run.go#Start
```
stopMonitor := startMonitoringAPI(...)

return asyncAbort, func(ctx context.Context) error {
  if err := stopMonitor() {
    return err
  }
  
  ...
}

```

Reply:
Great idea!